test/Mono.Linker.Tests.Cases/DataFlow/NullableAnnotations.cs

```diff
@@ -0,0 +1,289 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Helpers;
+using DAM = System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute;
+using DAMT = System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes;
+
+namespace Mono.Linker.Tests.Cases.DataFlow
+{
+	[ExpectedNoWarnings]
+	class NullableAnnotations
+	{
+		[Kept]
+		struct TestStruct
+		{
+			[Kept]
+			[KeptBackingField]
+			public string FirstName { [Kept] get; [Kept] set; }
+			[Kept]
+			[KeptBackingField]
+			public string LastName { [Kept] get; [Kept] set; }
+		}
+
+		[Kept]
+		struct TestStructWithRucMethod
+		{
+			[Kept]
+			[KeptBackingField]
+			public string FirstName { [Kept] get; [Kept] set; }
+			[Kept]
+			[KeptBackingField]
+			public string LastName { [Kept] get; [Kept] set; }
+
+			[Kept]
+			[KeptAttributeAttribute (typeof (RequiresUnreferencedCodeAttribute))]
+			[RequiresUnreferencedCode ("message")]
+			void MethodWithRuc () { }
+		}
+
+		[Kept]
+		class TestClass
+		{
+			[Kept]
+			public string FirstName { [Kept][ExpectBodyModified] get; [Kept][ExpectBodyModified] set; }
+			[Kept]
+			public string LastName { [Kept][ExpectBodyModified] get; [Kept][ExpectBodyModified] set; }
+			[Kept]
+			[KeptBackingField]
+			public static string StaticString { [Kept]get; [Kept]set; }
+
+			public void Method () { }
+		}
+
+		[Kept]
+		public static void Main ()
+		{
+			NullableOfAnnotatedGenericParameterRequiresPublicProperties<TestStruct> ();
+			Type hasProperties = ReturnUnderlyingTypeThatRequiresProperties<Nullable<TestStruct>> (new ());
+			hasProperties.RequiresPublicProperties ();
+			RequireMethodWithRUC ();
+
+			DamOnNullableKeepsUnderlyingMembers ();
+			UnderlyingTypeOfCreatedNullableOfAnnotatedTRequiresPublicProperties<TestStruct> ();
+			RequirePublicFieldsOnGenericParam<Nullable<StructWithFieldsReferencedThroughDamOnNullable>> ();
+			m<StructWithFieldsReferencedThroughDamOnNullable> ();
+
+			GenericClassTypeParamRequiresPublicFields<GenericClassTypeParam>.RunTests ();
+			GenericClassUnderlyingTypeParamRequiresPublicFields<Nullable<GenericClassTypeParam>>.RunTests ();
+
+			TestGetUnderlyingTypeOnClasses ();
+			TestGetUnderlyingTypeOnStructs ();
+			TestGetUnderlyingTypeOnNullableStructs ();
+			TestGetUnderlyingTypeOfCreatedNullableOnStructs ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2026", "message")]
+		static void RequireMethodWithRUC ()
+		{
+			var T = typeof (Nullable<TestStructWithRucMethod>);
```

Comment:
To test the analyzer we should also have a test which does just `typeof(Nullable<TestStructWithRucMethod>).RequiresAll()` - and that should also produce IL2026.